README
What is LiDE?
============
Linux Disk Extractor: The linux cat/dd command extended across networks. 
It helps you copy files across networks. 
You may copy a full disk, a partition or a file using this command.

Example usage
=============
    There is a disk called /dev/sda on ipaddress 192.168.1.1.
    There is another disk called /dev/sdb on ipaddress 192.168.1.2.
    Now to copy /dev/sda@192.168.1.1 to /dev/sda@192.168.1.2, 
    run the following command:
	On 192.168.1.1:
	    $ lide /dev/sda
	On 192.168.1.2:
	    netcat 192.168.1.1 31337 > /dev/sda && md5sum ReceivedFile.tst
    
    LiDE defaults to TCP/31337, but the port can be specified with -p.
	On 192.168.1.1:
	    $ lide -p 8888 FileToSend
	On 192.168.1.2:
	    netcat 192.168.1.1 8888 > ReceivedFile && md5sum ReceivedFile

To execute lide on a remote system from local PC, 
execute the following to make the file available on the network:
tar cf - ./lide | ssh forensics@192.168.1.1 'tar xf -; ./lide FileToSend'
***This will copy the file to the remote system and execute it.

To compile
==========
	make

To install
==========
	make install

   This would install /bin/lide. 
   You may edit the Makefile to whatever installation prefix as per your choice.

To uninstall
============
	make uninstall

To cleanup:
===========
	make clean


ChangeLog
LiDE "Linux Disk Extractor"

As of version 0.0.3.0 it has the same functionality as running a string of commands:
-dd to capture a file (more like cat though)
-pv "pipe viewer" to view the progress of the file transfer
-netcat to send the file over the network
-md5sum to report the unique file hash, in order to confirm a 100% good file transfer
-ifconfig to report the local IP address
All packaged into a small executable less than 19Kb in size.

The following command could accomplish the same results:
ifconfig | grep -A1 eth0 | grep inet | cut -f2 -d ":" | cut -f1 -d " " && md5sum File_to_Transfer && cat File_to_Transfer | pv | nc -l -p 31337


Summary ChangeLog

Original source code: 
https://github.com/joshualockhart/Hermes
by Josh Lockhart, joshualockhart@gmail.com, 30 Oct 2010

Fixed Original Source:
-Fixed hang, on end of file transfer.
-Submited Pull request to https://github.com/joshualockhart/Hermes

v0.0.1.1
-Added ChangeLog
-Renamed project to LiME
-Renamed files.
-Fixed version.
-Patched Makefile

v0.0.1.2
-Removed option for verbose, always verbose.
-Changed default port.
-Removed new line for progress status (verbose).

v0.0.1.3
-Progress status removed cursor (looks cleaner).

v0.0.1.4
-Re-enable cursor after program termination.

v0.0.1.5
-Fixed Help to reflect program as it is now layed out.
-Changed displayed wording.
-Added colors to output.
-Cleaned up Makefile

v0.0.2.0
-Added MD5 function for source file
MD5 Function:
http://stackoverflow.com/questions/1220046/how-to-get-the-md5-hash-of-a-file-in-c/1220177#1220177
D'Nabre, 2 Aug 2009

v0.0.2.1
-Fixed MD5 function error displayed if no file available.

v0.0.3.0
-Added get local IP Address of eth0 (hardcoded)
-Added display remote command to acquire file transfer
-Clean up comments 
Get local IP Address Function:
http://www.geekpage.jp/en/programming/linux-network/get-ipaddr.php
Akimichi Ogawa

v0.0.3.1
-Added option to choose interface, now can display remote command to acquire file transfer on interface other than eth0 

v0.0.3.2
-Changed display of executed program, easier to read.


Referenced Sources

https://github.com/joshualockhart/hermes
Thanks to Joshua Lockhart for getting my project off the ground. 

MD5 Function
http://stackoverflow.com/questions/1220046/how-to-get-the-md5-hash-of-a-file-in-c/1220177#1220177
D'Nabre, 2 Aug 2009

Get local IP Address Function:
http://www.geekpage.jp/en/programming/linux-network/get-ipaddr.php
Akimichi Ogawa


nwio.c
#include <stdio.h>        // C library to perform Input/Output operations
#include <unistd.h>       // C standard symbolic constants and types
#include <string.h>       // C Strings
#include <stdlib.h>       // C Standard General Utilities Library
#include <arpa/inet.h>    // C definitions for internet operations
#include <sys/socket.h>   // C Internet Protocol family
#include "LiDE.h"         //
//====Start-Local_IP_Address====
#include <sys/types.h>    // C data types                --|
#include <sys/ioctl.h>    // C generic I/O operation       |== Support IP Address Function
#include <netinet/in.h>   // C Internet Protocol family    |
#include <net/if.h>       // C Internet Protocol family  --|
//====End-Local_IP_Address====

/*
 * Send the file to the currently connected client, return the number of bytes sent or 0 on error. 
 * The moment there is an error with a send call, it just dies and returns -1,  
 * this could obviously be improved to make it a little "smarter".
 */

int send_file(FILE *fd, int connfd)
{
	int filesize,
	    number_complete_chunks,
	    final_chunk_size,
	    total_bytes_sent = 0,
	    bytes_sent = 0,
	    i;

	fseek(fd, 0L, SEEK_END);
	filesize = ftell(fd);
	rewind(fd);

	number_complete_chunks = filesize / CHUNK_SIZE;
	final_chunk_size = filesize % CHUNK_SIZE;

	char *chunk;

	if((chunk = (char *) malloc(CHUNK_SIZE)) == NULL)
	{
		perror("Malloc");
		exit(1);
	}


	for (i = 0; i < number_complete_chunks; i++)
	{
		fread(chunk, CHUNK_SIZE, 1, fd);

		if ((bytes_sent = send(connfd, chunk, CHUNK_SIZE, 0)) < 0)
			return -1;

		total_bytes_sent += bytes_sent;
		printf("\x1B[31m%d\x1B[32m of %d\040bytes\r\e[?25l\033[0m", total_bytes_sent, filesize); // the "\r" refreshes current line and "\e[?25l" removes the cursor
	}

	// send the last incomplete chunk at the end of the file.
	fread(chunk, final_chunk_size, 1, fd);

	if ((bytes_sent = send(connfd, chunk, final_chunk_size, 0)) < 0)
		return -1;

	total_bytes_sent += bytes_sent;
		printf("\x1B[32m%d of %d\040bytes\033[0m\n", total_bytes_sent, filesize);

	free(chunk);
	return total_bytes_sent;
}

/*
 * Wait for a port connection asking for our file.
 */

int listen_for_inbound_requests(int port, FILE *filename_fd)
{
	int listenfd, bytes_sent;
	int connfd;
	struct sockaddr_in serv, cli;	
	socklen_t len;

	//====Start-Local_IP_Address====
	int ipfd;
	struct ifreq ifr;
	ipfd = socket(AF_INET, SOCK_DGRAM, 0);
	ifr.ifr_addr.sa_family = AF_INET; /* I want to get an IPv4 IP address */
	strncpy(ifr.ifr_name, ip_interface, IFNAMSIZ-1); /* I want IP address attached to "eth0" */
	ioctl(ipfd, SIOCGIFADDR, &ifr);
	close(ipfd);
	//====End-Local_IP_Address====

	char *ip_buff; // Somewhere to put an IP address.
	
	if((ip_buff = (char *) malloc(INET_ADDRSTRLEN)) == NULL)
	{
		perror("Malloc");
		exit(1);
	}

	if ((listenfd = socket(AF_INET, SOCK_STREAM, 0)) == -1){
		perror("Error creating socket");
	}

	int tr = 1;
	if (setsockopt(listenfd,SOL_SOCKET,SO_REUSEADDR,&tr,sizeof(int)) == -1) { 
		perror("setsockopt");
		exit(1);
	}

	bzero(&serv, sizeof(serv));
	serv.sin_family = AF_INET;
	serv.sin_addr.s_addr = htonl(INADDR_ANY);
	serv.sin_port = port;

	if ((bind(listenfd, (struct sockaddr *) &serv, sizeof(serv))) == -1){
		perror("Bind error");
		exit(1);
	} 
	
	if ((listen(listenfd, 5)) == -1){ // We don't require a huge backlog (arg 2). 
		perror("Send error");
		exit(1);
	}

	//====Start-Local_IP_Address====
	printf("\x1B[32mUsing the IP from interface:\x1B[31m %s\033[0m\n", ip_interface);
	printf("\x1B[33mTo capture the sent file, enter the following on the remote system: \n\x1B[31mnetcat %s %d > received_file && md5sum received_file\033[0m\n", inet_ntoa(((struct sockaddr_in *)&ifr.ifr_addr)->sin_addr), ntohs(serv.sin_port)); /* display result */
	//====End-Local_IP_Address====

	printf("\x1B[32m\nSending on port %d...\033[0m\n", ntohs(serv.sin_port));

	for (;;) {
		len = sizeof(cli);
		connfd = accept(listenfd, (struct sockaddr *) &cli, &len);

		inet_ntop(AF_INET, &(cli.sin_addr), ip_buff, INET_ADDRSTRLEN);
		printf("\x1B[32mConnection from %s, port %d\033[0m\n", ip_buff, ntohs(cli.sin_port));

		if ((bytes_sent = send_file(filename_fd, connfd)) < 0)
			printf("\x1B[31mError transferring file.\033[0m\n\e[?25h");
		else
			printf("\x1B[32mFinished transfer to %s\n\%d bytes sent.\033[0m\n\e[?25h", ip_buff, bytes_sent);exit(0);
	}

	free(ip_buff);

	return 0;
}


mktarball.sh
#!/bin/bash

mkdir -p Source
cd Source

curSrcVer=$(ls -l | tr -s ' ' | cut -f9 -d " " | grep 'v[0-9]' | tail -n1)
preSrcVer=$(ls -l | tr -s ' ' | cut -f9 -d " " | grep 'v[0-9]' | tail -n2 | head -n1)
mkdir -p $curSrcVer/Patch
diff $preSrcVer/LiDE.c $curSrcVer/LiDE.c >$curSrcVer/Patch/LiDE.c.patch
diff $preSrcVer/LiDE.h $curSrcVer/LiDE.h >$curSrcVer/Patch/LiDE.h.patch
diff $preSrcVer/nwio.c $curSrcVer/nwio.c >$curSrcVer/Patch/nwio.c.patch
diff $preSrcVer/Makefile $curSrcVer/Makefile >$curSrcVer/Patch/Makefile.patch
diff $preSrcVer/README $curSrcVer/README.md >$curSrcVer/Patch/README.patch

rm -r Current
mkdir -p Current/Patch
cp $curSrcVer/* Current/ 2>/dev/null
diff Original\ Sources/hermes/hermes.c $curSrcVer/LiDE.c >Current/Patch/LiDE.c.patch
diff Original\ Sources/hermes/hermes.h $curSrcVer/LiDE.h >Current/Patch/LiDE.h.patch
diff Original\ Sources/hermes/nwio.c $curSrcVer/nwio.c >Current/Patch/nwio.c.patch
diff Original\ Sources/hermes/Makefile $curSrcVer/Makefile >Current/Patch/Makefile.patch
diff Original\ Sources/hermes/README $curSrcVer/README.md >Current/Patch/README.patch
diff Original\ Sources/hermes/mktarball.sh $curSrcVer/mktarball.sh >Current/Patch/mktarball.sh.patch

cd ..

#*****Start Configuration \ Variables*****
name="LiDE"    #Name of project.
version="`grep PROG_VERSION $name.h | cut -f2 -d \"\\"\"`"  #Replace everything between Parentheses with "`Your command to extract version of project`"
files="nwio.c $name.c $name.h Makefile ChangeLog.md README.md Source mktarball.sh" #List files\directories you want to include in your source code distribution.
#*****End Configuration \ Variables*****

echo "[$name] making tarball..."
mkdir $name-$version
tar --exclude=CVS --exclude-backup -cf temporaryplace-$version.tar $files
tar -C $name-$version -xf temporaryplace-$version.tar
rm -f $name-$version.tar.gz
tar -czf $name-$version.tar.gz --numeric-owner $name-$version
rm -rf $name-$version
rm -rf temporaryplace-$version.tar
chmod 0644 $name-$version.tar.gz
echo "[$name] done"


LiDE.c
#include <stdio.h>        // C library to perform Input/Output operations
#include <unistd.h>       // C standard symbolic constants and types
#include <string.h>       // C Strings
#include <stdlib.h>       // C Standard General Utilities Library
#include <arpa/inet.h>    // C definitions for internet operations
#include <sys/socket.h>   // C Internet Protocol family
#include <sys/types.h>    // C data types
#include "LiDE.h"         //    
//====Start-MD5-Support====
#include <sys/stat.h>     // C data returned by the stat() function    --|
#include <sys/mman.h>     // C memory management declarations            |== Support MD5 Function
#include <fcntl.h>        // C file control options                      |
#include <openssl/md5.h>  // C support MD5 functions                   --|

unsigned char result[MD5_DIGEST_LENGTH];

// Print the MD5 sum as hex-digits.
void print_md5_sum(unsigned char* md) {
    int i;
    for(i=0; i <MD5_DIGEST_LENGTH; i++) {
            printf("%02x",md[i]);
    }
}

// Get the size of the file by its file descriptor
unsigned long get_size_by_fd(int fd) {
    struct stat statbuf;
    if(fstat(fd, &statbuf) < 0) exit(-1);
    return statbuf.st_size;
}
//====End-MD5-Support====

void print_help(void){
	printf("\x1B[33m%s (\033[1mLi\033[22mnux \033[1mD\033[22misk \033[1mE\033[22mxtractor) - v%s\033[0m \n\n", PROG_NAME, PROG_VERSION);
	printf("This is a linux TCP enabled cat/dd utility. \n\
It enables you to copy a file, full disk or partition across network.\n\n\
Usage; \n\
$ \x1B[32m./lide [-h]\033[0m                        Prints this help.\n\n\
To host a file:\n\
$ \x1B[32m./lide <filename>\033[0m                  Will use default port 31337\n\
    or\n\
$ \x1B[32m./lide -p <port> <filename>\033[0m        To specify a port to send on.\n\
$ \x1B[32m./lide -i <eth0|wlan0> <filename>\033[0m  To specify an interface to send on.\n\
$ \x1B[32m./lide -i <eth0|wlan0> -p <port> <filename>\033[0m\n\n\
Then to retrieve it on another host:\n\
$ \x1B[32mnetcat [LiDE_Host_IP] [PORT] > retrieved_file && md5sum retrieved_file\033[0m\n\n");
	exit(0);
}

int main(int argc, char *argv[])
{
	FILE *filename_fd;
	char *filename_to_serve;
	int listen_port = htons(31337);

	//====Start-MD5-Support====
	int file_descript;
	unsigned long file_size;
	char *file_buffer;
	//====End-MD5-Support====
	//====Start-Local_IP_Address====
	ip_interface = "eth0";             // Default network interface
	//====End-Local_IP_Address====

	if (argc < 2)
		print_help();

	char c;
	while ((c = getopt(argc, argv, "hi:p:")) != -1) {
		switch(c){

			case 'h':
				print_help();
				break;
			//====Start-Local_IP_Address====
			case 'i':
				if (optarg != NULL){
					ip_interface = optarg;
					break;
				}
				else break;
			//====End-Local_IP_Address====
			case 'p':
				if ((optarg != NULL) && (optarg[0] != '\0')){

					listen_port = htons(atoi(optarg));

					if (listen_port < 1024 && getuid() > 0){
						printf("Invalid port number. Binding to a port lower than 1024 requires root priveleges.\n");
						exit(0);
					}
					else {
						listen_port = htons(atoi(optarg));
						break;
					}
				}
				else break;
			default:
				print_help();
				break;
		}
	}

	// Get the filename, which is the argument without an option.
	if ((argc - optind) == 1)
	{
		if (strlen(argv[argc-1]) > 255)
		{
			printf("Filename too long.\n");
			exit(0);
		}

		if((filename_to_serve = malloc(strlen(argv[argc-1]) + 1)) == NULL)
		{
			perror("Malloc");
			exit(1);
		}
		strncpy(filename_to_serve, argv[argc-1], strlen(argv[argc-1]));
	}

	if ((filename_fd = fopen(filename_to_serve, "r")) != NULL){
		//====Start-MD5-Functions==== 
		file_descript = open(filename_to_serve, O_RDONLY);
		file_size = get_size_by_fd(file_descript);
    
		file_buffer = mmap(0, file_size, PROT_READ, MAP_SHARED, file_descript, 0);
		MD5((unsigned char*) file_buffer, file_size, result);
		munmap(file_buffer, file_size); 

		printf("\x1B[33m%s (\033[1mLi\033[22mnux \033[1mD\033[22misk \033[1mE\033[22mxtractor) - v%s\033[0m \n\n", PROG_NAME, PROG_VERSION);
		printf("\x1B[32mMD5 Hash:  ");
		print_md5_sum(result);
		printf("\nFilename:  %s\nFile size: %lu\040bytes\033[0m\n\n", argv[argc-1], file_size);
		//====End-MD5-Functions====

		listen_for_inbound_requests(listen_port, filename_fd);		
		printf("%d\n", listen_port);
	}

	else
		perror("Error opening file");

	free(filename_to_serve);

	return 0;
}
